File: AdventOfCode2021/3/third.py
def get_data(filename):
    '''
    Read data from a file, put it in list and transform to integers.
    '''
    myfile = open(filename)
    return myfile.read().split()

def count_digits(data):
    '''
    Counts number of ones at each index for all elements of input data.
    '''
    data_length = len(data)/2
    digit_count = [0]*len(data[0])
    for numbers in data:
        for i,digit in enumerate(numbers):
            if digit == '1': digit_count[i] += 1
    # Returns list containing one or zero depending of which one occurs more at given index
    return list(map(lambda x: '1' if x >= data_length else '0' , digit_count))

def count_digit(data, index, oxygen):
    '''
    Calculates if one or zero occurs more times at given index that is passed in as argument.
    '''
    data_length = len(data)/2
    counter = 0
    for num in data:
        if num[index] == '1': counter += 1
    if oxygen:
        # Calculating oxygen return the number that occurs less
        return '1' if counter >= data_length else '0'
    else:
        # Calculating co2 return number that occurs most
        return '0' if counter >= data_length else '1'

def reduce_data(data, oxygen = True):
    '''
    Removes elements of data based on index of an element mathcing return from count_digit function.
    '''
    new_data = data
    index = 0
    while len(new_data) > 1:
        # Check which digit occurs more or less based on oxygen param.
        count = count_digit(new_data, index, oxygen)
        # Filter out elements of array that have value returnd from count_digit at given index.
        new_data = list(filter(lambda x: x[index] == count, new_data))
        index += 1
    return new_data

def puzzle_one(data):
    # List with one or zeroes, depeding on which one occurs more in all of the elements at each index.
    gamma = count_digits(data)
    # Epsilon in reverse of gamma. We switch 1s for 0os and vice versa.
    epsilon = list(map(lambda x: '0' if x == '1' else '1', gamma))
    #Convert both to decimal.
    dec_gamma = int(''.join(gamma),2)
    dec_epsilon = int(''.join(epsilon),2)
    return dec_gamma * dec_epsilon

def puzzle_two(data):
    # For each index in elemnts, check which of the values at that index occurs least, and remove them, 
    # until there is only one element left in list.
    oxygen_generator = int(reduce_data(data, True)[0], 2)
    # Similar to oxygen, only it removes elements at given index that have values that occurs the most.
    co2_scrubber = int(reduce_data(data, False)[0], 2)
    return oxygen_generator * co2_scrubber

print("Puzzle three, first part solution: ")
# If we run into error, call get_data function with full path to the file.
print(puzzle_one(get_data('input.txt')))
print("Puzzle three, first part solution: ")
# If we run into error, call get_data function with full path to the file.
print(puzzle_two(get_data('input.txt')))

File: AdventOfCode2021/3/third.js
const fs = require('fs');

function getData(filename){
    const file = fs.readFileSync(filename, 'utf-8')
    const input = file.split(/\r?\n/);
    return input;
}

function countDigits(inp){
    let inpLength = inp.length/2;
    let digitCounter = Array(inp[0].length).fill(0);
    inp.forEach(num => {
        for(let i = 0; i < num.length; i++){
            if (num[i] === '1') digitCounter[i]++;
        }
    })
    return digitCounter.map(num => num >= inpLength ? 1 : 0);
}

function countDigit(inp, index, oxygen){
    let inpLength = inp.length/2;
    let counter = 0;
    inp.forEach(num => {if(num[index] === '1') counter++;} )
    if (oxygen){
        return counter >= inpLength ? '1' : '0';
    } else{
        return counter >= inpLength ? '0' : '1';
    }
}

function reduceInp(inp, oxygen = true ){
    let input = inp;
    let index = 0;
    while(input.length > 1){
        let count = countDigit(input, index, oxygen);
        input = input.filter(num => num[index] === count)
        index++;
    }
    return input;
}

function puzzleOne(inp){
    const gamma = countDigits(inp);
    const epsilon = gamma.map(num => num === 1 ? 0: 1)
    const gammaRate = parseInt((gamma.join('')), 2);
    const epsilonRate = parseInt((epsilon.join('')), 2);
    return epsilonRate * gammaRate;
}

function puzzleTwo(inp) {
    const oxygenGenerator = parseInt((reduceInp(inp, true)[0]), 2)
    const co2Scrubber = parseInt((reduceInp(inp, false)[0]), 2)
    return oxygenGenerator * co2Scrubber;
}

//841526
console.log("Puzzle three, first part solution: ")
console.log(puzzleOne(getData('input.txt')))

//4790390
console.log("Puzzle three, second part solution: ")
console.log(puzzleTwo(getData('input.txt')));